BJS2-52615 : Стать ментором

lexsandr666 opened the pull request.

lexsandr666 committed:
BJS2-52615

lexsandr666 committed:
BJS2-52615 : Removing unnecessary exceptions

shakhzod99 commented on src/main/java/school/faang/user_service/service/MentorshipService.java:
```diff
@@ -0,0 +1,58 @@
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class MentorshipService {
```
Можно сделать MentorshipServiceImpl и создать MentorshipService интерфейс и поставить impl папку, чтобы по принципам Solid как говорится и в контроллере вызывать метод интерфейса как раз таки

lexsandr666 replied:
Поправил данный момент. Вынес реализацию в impl, и оставил интрефейс. Так же поправил в контроллере объекта с экземпляра класса на его интрефейс.

shakhzod99 commented on src/test/java/school/faang/user_service/controller/MentorshipControllerTest.java:
```diff
@@ -0,0 +1,200 @@
+
+    private long userId = 1L;
+    private long deleteUserId = 2L;
+    Long menteeFirstId = 2L;
```
эти  константы лучше сделать private static final полагаю и на самый верх ( по лекциям Влада)

lexsandr666 replied:
Не сказал бы что для тестов... это проблема =)
В целом конечно плохая реализация, сделал по приципу "копипаста", но лучше переписать. Увидел хорошую реалзиацию в PR Маши, она вынесли все параметры в отдлебный класс-генератор сущностей. И кода меньше в тестах и можно в любой момент расширить генератор объектов.

shakhzod99 commented on src/test/java/school/faang/user_service/service/MentorshipServiceTest.java:
```diff
@@ -0,0 +1,205 @@
+        Goal goalWithMentee = new Goal();
+        Goal goalWithoutMentee = new Goal();
+
+        mentor.setId(1L);
```
может всё вынести в констаты, чем Magic numbers/literals?)

lexsandr666 replied:
Чуть выше давал комментарий по аналогичной ситуации.

bortsovamaria commented on src/main/java/school/faang/user_service/service/MentorshipService.java:
```diff
@@ -0,0 +1,58 @@
+        if (isRemoved) {
+            userService.saveUser(mentor);
+        } else {
+            log.info("Mentor " + mentor.getUsername() + " does not have a mentee with "
```
Может еще стоит выбросить исключение, чтобы перехватил exception handler и отдал ответ не ОК, а что-то пошло не так)

lexsandr666 replied:
Тут двоякая ситуация. Вроде с одной стороны это вроде как и не исключение, если мы пытаемся удалить того чего нет(опять же смотря какая бизнес-логика), с другой стороны, хотелось бы преподставить информацию о том, что при попытке удалить запись мы её на нашли. Я конечно поправлю логику на выброс исключения, хотя у меня в данном случае есть вопросы =)

bortsovamaria commented on src/main/java/school/faang/user_service/service/MentorshipService.java:
```diff
@@ -0,0 +1,58 @@
+@RequiredArgsConstructor
+public class MentorshipService {
+    private final UserMapper userMapper;
+    private final UserService userService;
```
По ходу других ревью пришли к тому, что это лишний слой и можно вызвать нужный репо прям тут в сервисе

lexsandr666 replied:
Я оставлю сервис и постараюсь на встречи объяснить логику, почему при данных ситуациях лучше использовать сервис а не на прямую за данными лезть в репозиторий. Пока проект маленький, мы не заметим проблему, которая потом может "выстрелить".

bortsovamaria commented on src/test/java/school/faang/user_service/service/MentorshipServiceTest.java:
```diff
@@ -0,0 +1,205 @@
+    private UserService userService;
+
+    @Mock
+    private UserMapper userMapper;
```
Маппер не сложный, не имеет смысла его мокать, можно spy повесить

lexsandr666 replied:
Поправил.

bortsovamaria replied:
Остался Mock) А так вроде все ок

bortsovamaria commented on src/test/java/school/faang/user_service/service/MentorshipServiceTest.java:
```diff
@@ -0,0 +1,205 @@
+        UserDto dto1 = new UserDto();
+        UserDto dto2 = new UserDto();
+        List<User> mentees = List.of(mentee1, mentee2);
+
```
Как вариант можно вынести подготовку данных в отдельный метод или класс

lexsandr666 replied:
Полностью соглашусь. Увидел твой PR, подумал тоже переписать и вынести все в отдельный класс-генератор (билдер)

bortsovamaria commented:
Не видно изменений после комментариев... Например по мапперу в тесте.

lexsandr666 committed:
BJS2-55521 :
1) Rewrote services to interfaces and implementing classes
2) Moved data generator in tests to a separate generator class
3) Added exception throwing in deleteMentee and deleteMentor methods

miked4king commented on src/main/java/school/faang/user_service/service/impl/UserServiceImpl.java:
```diff
@@ -0,0 +1,26 @@
+import school.faang.user_service.repository.UserRepository;
+import school.faang.user_service.service.UserService;
+
+@Slf4j
```
вроде ничего не логгируется, можно убрать

miked4king commented on src/main/java/school/faang/user_service/service/impl/MentorshipServiceImpl.java:
```diff
@@ -0,0 +1,57 @@
+    private final UserMapper userMapper;
+    private final UserService userService;
+
+    public List<UserDto> getMentees(long userId) {
```
на реализованные методы интерфейса принято вешать аннотацию @Override, добавь пж на всех нужных методах. 
в другом сервисе тоже

miked4king commented on src/main/java/school/faang/user_service/service/impl/MentorshipServiceImpl.java:
```diff
@@ -0,0 +1,57 @@
+        boolean isRemoved = mentor.getMentees().removeIf(mentee -> mentee.getId().equals(menteeId));
+
+        if (!isRemoved) {
+            throw new EntityNotFoundException("Mentor " + mentor.getUsername() + " does not have a mentee with " + menteeId + " id");
```
Давай возьмем за правило вместо конкатенации использовать String.format()
так гораздо удобнее читать сообщение и будет соблюдаться одинаковый формат по всему проекту.
Видел в другом месте ты его использовал)

miked4king commented on src/main/java/school/faang/user_service/controller/MentorshipController.java:
```diff
@@ -0,0 +1,76 @@
+@RestController
+@Validated
+@Slf4j
+@RequestMapping("/mentorship")
```
версионирование ведь вынесено в yaml файл?

miked4king replied:
Саша привет, тут я непонятно написал)
мы с ребятами давно следим за версионированием API, и общую часть "/api/v1" мы выносим в файл application.yaml, и подтягиваем сюда через плэйсхолдер, приведи пожалуйста этот эндпоинт к общему виду.

Если не понятно можешь посмотреть у ребят ПРы

miked4king reviewed (changes requested):
Отличная работа, молодец.
Чистый код, приятно читать

lexsandr666 committed:
BJS2-55521 :
1) Fix
